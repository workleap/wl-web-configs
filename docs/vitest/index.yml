order: 110
expanded: true
label: Vitest
