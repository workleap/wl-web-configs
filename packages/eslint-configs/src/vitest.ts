import vitestPlugin from "@vitest/eslint-plugin";
import type { Linter } from "eslint";
import type { ConfigWithExtends } from "./types.ts";

export interface VitestConfigOptions {
    enabled?: boolean;
    rules?: Partial<Linter.RulesRecord>;
}

export const vitestGlobalIgnores = [];

export function vitestConfig(options: VitestConfigOptions = {}) {
    const {
        enabled = true,
        rules = {}
    } = options;

    if (!enabled) {
        return [];
    }

    const config: ConfigWithExtends[] = [{
        name: "@workleap/eslint-configs/vitest",
        files: [
            "**/*.test.{js,jsx,ts,tsx}",
            "**/*-test.{js,jsx,ts,tsx}",
            "**/__tests__/*.{js,jsx,ts,tsx}",
            "**/test.{js,jsx,ts,tsx}"
        ],
        plugins: {
            // @ts-expect-error temporary code until defineConfig is supported.
            vitest: vitestPlugin
        },
        // Waiting for defineConfig support: https://github.com/vitest-dev/eslint-plugin-vitest/issues/771
        // extends: [
        //     vitestPlugin.configs.recommended
        // ],
        rules: {
            ...vitestPlugin.configs.recommended.rules,
            "vitest/no-commented-out-tests": "off",
            // Positioned last to allow the consumer to override any rules.
            ...rules
        }
    }];

    return config;
};
