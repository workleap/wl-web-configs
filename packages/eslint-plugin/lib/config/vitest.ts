import { reactTestFiles, testFiles } from "../utils/patterns";

import type { Linter } from "eslint";

const config: Linter.Config = {
    overrides: [
        {
            files: [...testFiles, ...reactTestFiles],
            plugins: ["@vitest"],
            extends: ["plugin:@vitest/legacy-recommended"]
        }
    ]
};

// Using TypeScript "export" keyword until ESLint support ESM.
// Otherwise we must deal with a weird CommonJS output from esbuild which is not worth it.
// For more info, see: https://github.com/evanw/esbuild/issues/1079
export = config;

