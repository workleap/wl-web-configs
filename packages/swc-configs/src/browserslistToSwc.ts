// Inspired by https://github.com/marcofugaro/browserslist-to-esbuild/blob/main/src/index.js.
//
// This function exist because browserslist-rs (which is used under the hood by SWC) doesn't support browserslist "extends" syntax: https://github.com/browserslist/browserslist-rs#limitations.
// Therefore, we must parse ourself the ".browserslistrc" file to extract the browser entries and transform them into valid SWC "env.targets".
//
// The function will convert an array of browserslist entries from:
//
// ['and_chr 115',   'and_ff 115',   'and_qq 13.1',
// 'and_uc 15.5',   'android 115',  'baidu 13.18',
// 'bb 10',         'bb 7',         'chrome 116',
// 'chrome 115',    'chrome 114',   'chrome 113',
// 'chrome 112',    'chrome 111',   'chrome 110',
// 'chrome 109',    'chrome 108',   'chrome 107',
// 'edge 115',      'edge 114',     'edge 113',
// 'edge 112',      'edge 111',     'edge 110',
// 'edge 109',      'edge 108',     'edge 107',
// 'edge 106',      'firefox 116',  'firefox 115',
// 'firefox 114',   'firefox 113',  'firefox 112',
// 'firefox 111',   'firefox 110',  'firefox 109',
// 'firefox 108',   'firefox 107',  'ie 11',
// 'ie 10',         'ie 9',         'ie 8',
// 'ie 7',          'ie 6',         'ie 5.5',
// 'ie_mob 11',     'ie_mob 10',    'ios_saf 16.5',
// 'ios_saf 16.4',  'ios_saf 16.3', 'ios_saf 16.2',
// 'ios_saf 16.1',  'ios_saf 16.0', 'ios_saf 15.6-15.7',
// 'ios_saf 15.5',  'ios_saf 15.4', 'ios_saf 15.2-15.3',
// 'kaios 3.0-3.1', 'kaios 2.5',    'op_mini all',
// 'op_mob 73',     'opera 101',    'opera 100',
// 'opera 99',      'opera 98',     'opera 97',
// 'opera 96',      'opera 95',     'opera 94',
// 'opera 93',      'opera 92',     'safari 16.5',
// 'safari 16.4',   'safari 16.3',  'safari 16.2',
// 'safari 16.1',   'safari 16.0',  'safari 15.6',
// 'safari 15.5',   'safari 15.4',  'safari 15.2-15.3',
// 'samsung 21',    'samsung 20',   'samsung 19.0',
// 'samsung 18.0',  'samsung 17.0', 'samsung 16.0',
// 'samsung 15.0',  'samsung 14.0', 'samsung 13.0',
// 'samsung 12.0']
//
// To an object literal of SWC "env.targets":
//
// {
//  chrome: '107',
//  firefox: '107',
//  android: '115',
//  edge: '106',
//  ie: '5.5',
//  opera: '73',
//  safari: '15.2',
//  samsung: '12'
// }

import browserslist from "browserslist";

// Original browserslist-rs supported browsers (https://github.com/browserslist/browserslist-rs/blob/99a3244fc8c0e631a80a9cae5c41dca6c5a2aae5/build.rs#L10)
// minus the browsers that are discarted by SWC (https://github.com/swc-project/swc/blob/main/crates/preset_env_base/src/lib.rs#L105).
const SwcSupportedBrowsers = [
    "ie",
    "edge",
    "firefox",
    "chrome",
    "safari",
    "opera",
    "ios",
    "android",
    "op_mob",
    "and_chr",
    "and_ff",
    "ie_mob",
    "samsung"
] as const;

// SWC browsers aliases (https://github.com/swc-project/swc/blob/main/crates/preset_env_base/src/lib.rs#L90).
const SwcMobileBrowserAliases: Record<string, string> = {
    "and_chr": "chrome",
    "and_ff": "firefox",
    "ie_mob": "ie",
    "ios_saf": "ios",
    "op_mob": "opera"
};

function parseBrowserslistEntry(entry: string) {
    // "chrome 11" --> ["chrome", "11"]
    const values = entry.split(" ");

    let browser = values[0];
    let version = values[1];

    // "and_chr" --> "chrome"
    if (SwcMobileBrowserAliases[browser]) {
        browser = SwcMobileBrowserAliases[browser];
    }

    // "11.0-12.0" --> "11.0"
    if (version.includes("-")) {
        version = version.slice(0, version.indexOf("-"));
    }

    // "11.0" --> "11"
    if (version.endsWith(".0")) {
        version = version.slice(0, -2);
    }

    return {
        browser,
        version
    };
}

export function createSwcTargetsFromBrowserslistEntries(entries: string[]) {
    return entries.reduce((acc, x: string) => {
        const { browser, version } = parseBrowserslistEntry(x);

        // Exclude browsers that are not supported by SWC.
        if (!((SwcSupportedBrowsers as unknown) as string[]).includes(browser)) {
            return acc;
        }

        // Let's make TS happy as we are now certain that the browser is part
        // of the supported browsers list.
        const _browser = browser as keyof typeof SwcSupportedBrowsers;

        // SWC "env.targets" only support a single entry per target (browser).
        // Therefore, we loop through the browser versions to keep only the oldest one.
        if (acc[_browser]) {
            if (parseFloat(acc[_browser]) > parseFloat(version)) {
                acc[_browser] = version;
            }
        } else {
            acc[_browser] = version;
        }

        return acc;
    }, {} as Record<keyof typeof SwcSupportedBrowsers, string>);
}

export interface BrowserslistToSwcOptions extends Omit<browserslist.Options, "path"> {
    queries?: string | string[];
}

export function browserslistToSwc(options: BrowserslistToSwcOptions = {}) {
    const {
        queries,
        ...browserlistsOptions
    } = options;

    // Will return the entries matching the "queries" prop when provided, otherwise,
    // the function will load the closest ".browserslistrc" file.
    const entries = browserslist(queries, {
        path: "./browserslistrc",
        ...browserlistsOptions
    });

    return createSwcTargetsFromBrowserslistEntries(entries);
}
