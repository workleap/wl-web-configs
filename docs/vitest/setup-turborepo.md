---
order: 70
label: Setup with Turborepo
meta:
    title: Setup with Turborepo - Vitest
toc:
    depth: 2-3
---

# Setup with Turborepo

To configure [Vitest](https://vitest.dev/) in a monorepo managed with [Turborepo](https://turborepo.com/), follow these steps ðŸ‘‡

## Setup the workspace

### Install the packages

Open a terminal at the root of the solution's workspace (the root of the repository) and install the following packages:

```bash
pnpm add -D turbo
```

### Configure Turborepo

First, create a configuration file named `turbo.json` at the root of the solution's workspace:

``` !#8
workspace
â”œâ”€â”€ packages
â”œâ”€â”€â”€â”€ pkg-1
â”œâ”€â”€â”€â”€â”€â”€ src
â”œâ”€â”€â”€â”€â”€â”€â”€â”€ ...
â”œâ”€â”€â”€â”€â”€â”€ package.json
â”œâ”€â”€ package.json
â”œâ”€â”€ turbo.json
```

Then, open the newly created file and copy/paste the following content:

```json !#5-7
{
    "$schema": "https://turbo.build/schema.json",
    "ui": "tui",
    "tasks": {
        "test": {
            "outputs": ["node_modules/.cache/vitest/**"]
        }
    }
}
```

### Add a CLI script

Finally, add the following script to your solution's workspace `package.json` file:

``` !#7
workspace
â”œâ”€â”€ packages
â”œâ”€â”€â”€â”€ pkg-1
â”œâ”€â”€â”€â”€â”€â”€ src
â”œâ”€â”€â”€â”€â”€â”€â”€â”€ ...
â”œâ”€â”€â”€â”€â”€â”€ package.json
â”œâ”€â”€ package.json    <------- (this one!)
â”œâ”€â”€ turbo.json
```

```json package.json
{
    "test": "turbo run test --continue"
}
```

## Setup a project

### Install the packages

Open a terminal at the root of the project (`packages/pkg-1` for this example) and install the following packages:

```bash
pnpm add -D vitest
```

### Configure Vitest

!!!info
`workleap/web-configs` does not offer any custom configuration for Vitest.
!!!

First, create a configuration file named `vitest.config.ts` at the root of the project:

``` !#7
workspace
â”œâ”€â”€ packages
â”œâ”€â”€â”€â”€ pkg-1
â”œâ”€â”€â”€â”€â”€â”€ src
â”œâ”€â”€â”€â”€â”€â”€â”€â”€ ...
â”œâ”€â”€â”€â”€â”€â”€ package.json
â”œâ”€â”€â”€â”€â”€â”€ vitest.config.ts
â”œâ”€â”€ package.json
â”œâ”€â”€ turbo.json
```

Then, open the newly created file and copy/paste the following content:

```ts packages/pkg-1/vitest.config.ts
import { defineConfig } from "vitest/config";

export default defineConfig({
    test: {
        include: ["tests/**/*.test.ts"],
        exclude: ["node_modules", "dist"],
        reporters: "verbose"
    },
    cacheDir: "./node_modules/.cache/vitest"
});
```

!!!tip
For additional information about Vitest, refer to the [Vitest documentation](https://vitest.dev/). 
!!!

### Add a CLI script

Finally, add the following `test` script to your project `package.json` file. This script will be executed by Turborepo:

```json packages/pkg-1/package.json
{
    "test": "vitest --config vitest.config.ts --no-watch"
}
```

## Try it :rocket:

To test your new setup, add a test to the `packages/pkg-1` project:

```ts packages/pkg-1/tests/foo.test.ts
import { test } from "vitest";

test("will always pass", ({ expect }) => {
    expect(true).toBeTruthy();
});
```

Open a terminal at the root of the solution and execute the CLI script added earlier:

```bash
pnpm test
```

The terminal output should indicate that the test suite completed successfully.


