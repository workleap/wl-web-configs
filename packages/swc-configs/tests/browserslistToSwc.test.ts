import { describe, test } from "vitest";
import { browserslistToSwc, createSwcTargetsFromBrowserslistEntries } from "../src/browserslistToSwc.ts";

describe("createSwcTargetsFromBrowserslistEntries", () => {
    test.concurrent("when a browser is not supported, do not include a target for the browser", ({ expect }) => {
        const result = createSwcTargetsFromBrowserslistEntries([
            "and_uc 15.5",
            "baidu 13.18",
            "firefox 115"
        ]);

        expect(result).toMatchSnapshot();
    });

    test.concurrent("when a browser is a mobile alias to a desktop browser, include a target for the desktop browser", ({ expect }) => {
        const result = createSwcTargetsFromBrowserslistEntries([
            "and_chr 115",
            "and_ff 115",
            "ie_mob 11",
            "ios_saf 16.5",
            "op_mob 73"
        ]);

        expect(result).toMatchSnapshot();
    });

    test.concurrent("when there are multiple versions of the same browser, only include the oldest version of the browser", ({ expect }) => {
        const result = createSwcTargetsFromBrowserslistEntries([
            "android 115",
            "chrome 115",
            "chrome 113",
            "chrome 108",
            "chrome 112",
            "chrome 107",
            "opera 101",
            "chrome 111",
            "chrome 110",
            "chrome 109",
            "chrome 114",
            "edge 115",
            "edge 114",
            "edge 113",
            "firefox 109",
            "firefox 116",
            "firefox 115",
            "firefox 107",
            "firefox 114",
            "firefox 113",
            "firefox 108",
            "firefox 112",
            "firefox 111",
            "firefox 110"
        ]);

        expect(result).toMatchSnapshot();
    });

    test.concurrent("when a browser is a mobile alias for a desktop browser, and the mobile alias version is older than the oldest desktop browser version, use the alias version", ({ expect }) => {
        const result = createSwcTargetsFromBrowserslistEntries([
            "and_chr 115",
            "chrome 116"
        ]);

        expect(result).toMatchSnapshot();
    });

    test.concurrent("when a browser version is a range, only keep the lower value of the range", ({ expect }) => {
        const result = createSwcTargetsFromBrowserslistEntries([
            "safari 15.2-15.3"
        ]);

        expect(result).toMatchSnapshot();
    });

    test.concurrent("when a browser version has a 0 digit, remove the digit", ({ expect }) => {
        const result = createSwcTargetsFromBrowserslistEntries([
            "samsung 19.0"
        ]);

        expect(result).toMatchSnapshot();
    });

    test.concurrent("when a browser version has an non 0 digit, keep the digit", ({ expect }) => {
        const result = createSwcTargetsFromBrowserslistEntries([
            "safari 15.4"
        ]);

        expect(result).toMatchSnapshot();
    });

    test.concurrent("when a browser version is \"all\", keep \"all\"", ({ expect }) => {
        const result = createSwcTargetsFromBrowserslistEntries([
            "chrome all"
        ]);

        expect(result).toMatchSnapshot();
    });
});

test.concurrent("when queries are provided, use the queries", ({ expect }) => {
    const result = browserslistToSwc({
        queries: ["IE 11", "chrome 113", "firefox 107"]
    });

    expect(result).toMatchSnapshot();
});
