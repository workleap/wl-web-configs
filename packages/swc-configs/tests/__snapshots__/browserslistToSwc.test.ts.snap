// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createSwcTargetsFromBrowserslistEntries > when a browser is a mobile alias for a desktop browser, and the mobile alias version is older than the oldest desktop browser version, use the alias version 1`] = `
{
  "chrome": "115",
}
`;

exports[`createSwcTargetsFromBrowserslistEntries > when a browser is a mobile alias to a desktop browser, include a target for the desktop browser 1`] = `
{
  "chrome": "115",
  "firefox": "115",
  "ie": "11",
  "ios": "16.5",
  "opera": "73",
}
`;

exports[`createSwcTargetsFromBrowserslistEntries > when a browser is not supported, do not include a target for the browser 1`] = `
{
  "firefox": "115",
}
`;

exports[`createSwcTargetsFromBrowserslistEntries > when a browser version has a 0 digit, remove the digit 1`] = `
{
  "samsung": "19",
}
`;

exports[`createSwcTargetsFromBrowserslistEntries > when a browser version has an non 0 digit, keep the digit 1`] = `
{
  "safari": "15.4",
}
`;

exports[`createSwcTargetsFromBrowserslistEntries > when a browser version is "all", keep "all" 1`] = `
{
  "chrome": "all",
}
`;

exports[`createSwcTargetsFromBrowserslistEntries > when a browser version is a range, only keep the lower value of the range 1`] = `
{
  "safari": "15.2",
}
`;

exports[`createSwcTargetsFromBrowserslistEntries > when there are multiple versions of the same browser, only include the oldest version of the browser 1`] = `
{
  "android": "115",
  "chrome": "107",
  "edge": "113",
  "firefox": "107",
  "opera": "101",
}
`;

exports[`when queries are provided, use the queries 1`] = `
{
  "chrome": "113",
  "firefox": "107",
  "ie": "11",
}
`;
